EIPS/eip-2537.md

```diff
@@ -0,0 +1,296 @@
+
+### Field to curve mapping
+
+Set of parameters for SWU mapping method is provided by [IETF](https://github.com/cfrg/draft-irtf-cfrg-hash-to-curve/blob/master/draft-irtf-cfrg-hash-to-curve.md#suites-for-bls12-381-suites-bls12381). 
+
+One should pay particular attention to the following fine points during implementation:
+- SWU method [itself](https://github.com/cfrg/draft-irtf-cfrg-hash-to-curve/blob/master/draft-irtf-cfrg-hash-to-curve.md#simplified-swu-for-ab--0-simple-swu-ab0)
+- [Cofactor cleanup](https://github.com/cfrg/draft-irtf-cfrg-hash-to-curve/blob/master/draft-irtf-cfrg-hash-to-curve.md#clearing-the-cofactor-cofactor-clearing)
+
+Once again, [hash to field](https://github.com/cfrg/draft-irtf-cfrg-hash-to-curve/blob/master/draft-irtf-cfrg-hash-to-curve.md#hashing-to-a-finite-field-hashtofield) is NOT a part of this EIP as it can be implemented in EVM and with different strategies.
+
+## Test Cases
+<!--Test cases for an implementation are mandatory for EIPs that are affecting consensus changes. Other EIPs can choose to include links to test cases if applicable.-->
+
+Due to the large test parameters space we first provide properties that various operations must satisfy. We use additive notation for point operations, capital letters (`P`, `Q`) for points, small letters (`a`, `b`) for scalars. Generator for G1 is labeled as `G`, generator for G2 is labeled as `H`, otherwise we assume random point on a curve in a correct subgroup. `0` means either scalar zero or point of infinity. `1` means either scalar one or multiplicative identity. `group_order` is a main subgroup order. `e(P, Q)` means pairing operation where `P` is in G1, `Q` is in G2. 
+
+Requeired properties for basic ops (add/multiply):
+
+- Commutativity: `P + Q = Q + P`
+- Additive negation: `P + (-P) = 0`
+- Doubling `P + P = 2*P`
+- Subgroup check: `group_order * P = 0`
+- Trivial multiplication check: `1 * P = P`
+- Multiplication by zero: `0 * P = 0`
+- Multiplication by the unnormalized scalar `(scalar + group_order) * P = scalar * P`
+
+Required properties for pairing operation:
+- Degeneracy `e(P, 0*Q) = e(0*P, Q) = 1` 
+- Bilinearity `e(a*P, b*Q) = e(a*b*P, Q) = e(P, a*b*Q)` (internal test, not visible through ABI)
+
+Test vector for all operations are expanded in this `csv` files in [repo](https://github.com/matter-labs/eip1962/tree/master/src/test/test_vectors/eip2537).
+
```

Comment:
There are no test vectors for BLS12_PAIRING.  There are two test vectors for BLS12_MAP_FIELD_TO_CURVE, one for G1 and one for G2.

Reply:
Mapping vectors are separate cause even while the spec has a single "mapping" function it's actually two different routines under it depending on the input length. I'll add pairing vectors today, totally forgot about them

Reply:
We started fuzzing the BLS implementation in geth (go vs. asm) yesterday and found some bugs after adding these test cases to the corpus: https://github.com/zkcrypto/bls12_381/tree/master/src/tests. @shamatar maybe you can add them to the matter-labs testvectors too.

Reply:
@MariusVanDerWijden Hey! Are you sure that the link is correct? .dat files look to be used to test point encoding and decompression: encoding is different in this ABI and decompression is just beyond the functionality of this EIP

Reply:
Sorry I meant I added the uncompressed points to our testing library which revealed a bug within `BLS12_MAP_FP_TO_G1` and `BLS12_MAP_FP2_TO_G2`. E.g. this input `0000000000000000000000000000000000000000000000000000000000000000000000000000000215a2313a0241779006bc033a776aeedae5de5ea6ee5a9b9` revealed a difference between the two implementations I tested.

Reply:
Do you mean you have added uncompressed points as corpus items to the fuzzer? Otherwise I don't understand how adding points as inputs lead to changes in a function that takes field elements as inputs

Reply:
Yes thats what I meant.

Reply:
Ok, I'll add it into my fuzzing corpus too (it's anyway based on the test vectors)

Reply:
I'm adding the specific vectors that broke our stuff too